Handle TimeStamp columns by converting int64 to time.Time and vice versa

echarlus opened the pull request:
Since one cannot yet replace the driver.DefaultParameterConverter to properly handle type conversions (see https://github.com/golang/go/issues/18415)
I've made a quick patch to automatically handle TimeStamp (int64) columns conversion to time.Time by using the column type provided by Crate within the query’s results and to convert back to int64 any time.

echarlus committed:
Since one cannot replace the driver.DefaultParameterConverter to properly handle type conversions, handle at least time.Time columns properly by using the column type provided by Crate within the query’s results. driver.DefaultParameterConverter override is scheduled for GO 1.9 (see https://github.com/golang/go/issues/18415)

herenow commented:
If we changed the output/parsing of time columns from int64 to time.Time, it would cause breaking changes.

Although it would be nice to parse time.Time to int64 when receiving arguments.

Let me give this some thought.

echarlus commented:
Hello,
the code I provided also converts time.Time found in query arguments to int64 so that Time objects can be transparently used by the upper layers. 
I'm not sure I'm following what you mean by "If we changed the output/parsing of time columns from int64 to time.Time, it would cause breaking changes." could you please elaborate ?
Thanks !

herenow commented:
Wouldn't this change the value we are assigning to each row column?

https://golang.org/pkg/database/sql/#Rows.Scan

So if we did:

```golang
// create table users ( created_at timestmap );
rows, err := db.Query("select created_at from users limit 1")
for rows.Next() {
	var created_at time.Time

	rows.Scan(&created_at)
}
``

From what I understood, this PR changes what we are scanning into the destination we pass into `rows.Scan()`, so if the destination is an `int64`, it would fail when we tried to assign a `time.Time` into it, wouldn't it?

Maybe I'm confused, and got this all wrong, I haven't worked on this for a while.

I'll give this code a test.

echarlus commented:
Since the conversion from int64 to time.Time is performed in the query method (which is called by Query) the row type would indeed by time.Time and your sample should work as expected. So I believe that my change does not break the Scan function. Let me know what your findings show ...
Thanks !

echarlus committed:
Now that GO 1.9 is released properly handle data types in the driver  only, no more tweaks in gorm

echarlus committed:
Improved GeoPoint by using a struct rather than an array of floats.
Fixed code according to the GitHub boat comments
Added Scan method on GeoPoint to allow gorm to handle GeoPoint embedded in an object

echarlus committed:
Handle Crate's Array column type

echarlus committed:
Use our own basic json encoder rather than the GO lib one to properly handle floats with null decimals to prevent creating an int typed column in crate when a float is required

echarlus committed:
Convert v to Float before formatting with %f format

echarlus committed:
Implemented encoding of arrays of map/arrays

echarlus committed:
Handle interface types in Array & Map

echarlus committed:
Fix to satisty Error interface

echarlus committed:
Fixed JSON encoding causing issue on crate 2.3.2 because a String was sent as JSON. Now use Marshaller interface to properly handle our own types

echarlus committed:
Merge branch 'time_column_handling' of https://github.com/echarlus/go-crate into time_column_handling

merge from head.

echarlus committed:
//Prevents rounding errors seen with floats like 0.01*41 which is 0.41000000000000003 ...
//See https://floating-point-gui.de/
//Float values are now truncated to 10^-6 which is more than enough for our needs

echarlus committed:
Complete fix for (Fix had only been applied to float in arrays, not in maps):
Prevents rounding errors seen with floats like 0.01*41 which is 0.41000000000000003 ...
See https://floating-point-gui.de/
Float values are now truncated to 10^-6 which is more than enough for our needs

echarlus committed:
Prevent crash when processing nil maps.
Speed up float processing by removing un-necessary conversion & comparison

echarlus committed:
Ticket SDRMDPTH-15 : Fixed string encoding in encodeArray.

echarlus committed:
Prepare for Crate 4.x : support authentication

echarlus committed:
Fixed map of map element handling and added support for Int into map of map.

echarlus committed:
return a CrateErr rather than &CrateErr to allow errors.As to be used; reformatted comments
